### PR TITLE
feat(attachments): port rehydrateAttachment hook (#52)

### PR DESCRIPTION
**Pull request by patrick-chinchill**
## Summary

- Port upstream `Adapter.rehydrateAttachment` hook + `Thread` serialization integration (issue #52).
- `Attachment` gains a serializable `fetch_metadata: dict[str, str] | None` field that survives JSON roundtrips and carries adapter-specific identifiers (Slack `url`+`teamId`, Teams `url`, Google Chat `resourceName`+`url`, Telegram `fileId`, WhatsApp `mediaId`). Persisted through `Message.to_json` / `from_json` / `from_json_compat`.
- `Adapter` protocol exposes an optional `rehydrate_attachment(attachment) -> Attachment` hook (no-op default on `BaseAdapter`). `Chat._rehydrate_message(raw, adapter=None)` now threads the active adapter and calls the hook on every attachment that lost its `fetch_data` closure during the queue/debounce JSON roundtrip — matches upstream `adapter?.rehydrateAttachment?.(att)` via duck-typing so adapters without the hook (e.g. `MockAdapter`) remain no-ops.
- Per-adapter implementations land on Slack, Teams, Google Chat, Telegram, WhatsApp. Discord / GitHub / Linear intentionally do **not** implement it (upstream parity — those platforms either use public URLs or have no file attachments).

## Fidelity impact

Closes all 3 `[concurrency: queue attachment rehydration]` gaps in `chat.test.ts`:

- `should call rehydrateAttachment on deserialized attachments missing fetchData`
- `should skip rehydration for attachments that already have fetchData`
- `should leave attachments unchanged when adapter has no rehydrateAttachment`

`verify_test_fidelity.py`: 40 missing → 37 missing (no new gaps introduced).

## Test plan

- [x] `uv run ruff check src/ tests/ scripts/`
- [x] `uv run ruff format --check src/ tests/ scripts/`
- [x] `uv run python scripts/audit_test_quality.py` (0 hard failures)
- [x] `uv run python scripts/verify_test_fidelity.py` (3 `[concurrency: queue attachment rehydration]` gaps closed; no new gaps)
- [x] `uv run pytest tests/ --tb=short -q` — 3564 passed, 2 skipped (was 3545 passed)
- [x] Slack multi-workspace rehydrate resolves per-team token via `get_installation`
- [x] JSON-roundtrip fidelity simulated via a test helper that wraps `state.enqueue` with `json.loads(json.dumps(entry.message.to_json()))` (mirrors upstream's `vi.mocked(state.enqueue).mockImplementation`).

## Notes

- No divergences from upstream in the hook protocol or call semantics. The Python default `BaseAdapter.rehydrate_attachment` is a no-op (returns the attachment unchanged); adapters that want platform-specific rehydration override it. `MockAdapter` does not inherit `BaseAdapter` and therefore lacks the attribute by default, which preserves the TS behavior that tests rely on ("should leave attachments unchanged when adapter has no rehydrateAttachment").
- Queued under `## Unreleased` in `CHANGELOG.md` for bundled `0.4.26.2`.
- Touches `src/chat_sdk/types.py` (+37 lines, mostly near `Attachment` dataclass L249 and `SerializedAttachment` L303) and `src/chat_sdk/chat.py` (+95 lines, `_rehydrate_message` at L2041 plus `_coerce_attachments` helper). Kept edits surgical to minimize rebase pain with other in-flight PRs (#50, #58, #59, #61).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Attachments now persist adapter-specific metadata so downloadable media are restored after queue/state roundtrips; per-adapter URL allowlists and workspace-specific token resolution protect and enable secure downloads across Google Chat, Slack, Teams, Telegram, and WhatsApp.

* **Tests**
  * Added extensive tests covering attachment rehydration, URL allowlisting, workspace-token behavior, and serialization edge cases.

* **Documentation**
  * Updated docs to describe serialization/recovery behavior and host allowlists.

### CODE REVIEW

**Comment by patrick-chinchill** [comment]
Review verdict: LGTM. Refreshed and reviewed latest head c871b2b against main. No issues found in attachment rehydration or provider download safety paths. Verification: targeted adapter/chat/type coverage passed (349 passed, 2 skipped) and ruff passed on touched files. Formal GitHub approval is blocked because the authenticated account owns this PR.

File: src/chat_sdk/chat.py [hunk 1]
```diff
@@ -2037,52 +2038,75 @@ def _detect_mention(self, adapter: Adapter, message: Message) -> bool:
...
+        callables on attachments have been stripped.  If ``adapter``
+        exposes a ``rehydrate_attachment`` hook we call it on every
+        attachment that lost its ``fetch_data`` closure so downstream
+        handlers can still download bytes.
+        """
+        # Matches upstream: if the entry is already a Message instance, its
+        # fetch_data closures never went through a JSON roundtrip, so we
+        # return it untouched — no rehydrate pass.
         if isinstance(raw, Message):
             return raw
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/chat.py]: Addressed in 26a08bd — the P1 issue was a real Python-specific divergence: our Redis/Postgres `dequeue` returns `Message` instances (upstream returns raw JSON dicts), so upstream's early-return on `Message` silently skipped rehydration for every persistent-backend dequeue. Fix: removed the early-return, fall through to rehydrate pass. Documented in docs/UPSTREAM_SYNC.md.

File: src/chat_sdk/adapters/teams/adapter.py [hunk 2]
```diff
@@ -580,11 +581,52 @@ def _create_attachment(self, att: dict[str, Any]) -> Attachment:
...
+
+        async def fetch_data() -> bytes:
+            import httpx
+
+            async with httpx.AsyncClient() as http:
+                resp = await http.get(url)
+                resp.raise_for_status()
+                return resp.content
+
+        return fetch_data
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/teams/adapter.py:662]: Addressed in 26a08bd — Teams `_build_teams_fetch_data` now uses the pooled `_get_http_session()` helper (matches the pattern at teams/adapter.py:1268).

File: src/chat_sdk/adapters/slack/adapter.py [hunk 3]
```diff
@@ -1763,6 +1770,71 @@ async def fetch_data() -> bytes:
...
+        async with httpx.AsyncClient() as http:
+            resp = await http.get(url, headers={"Authorization": f"Bearer {token}"})
+            resp.raise_for_status()
+            content_type = resp.headers.get("content-type", "")
+            if "text/html" in content_type:
+                raise RuntimeError(
+                    "Failed to download file from Slack: received HTML login page. "
+                    'Ensure your Slack app has the "files:read" OAuth scope.'
+                )
+            return resp.content
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/slack/adapter.py:1941]: Deferred — Slack adapter has no pooled aiohttp helper (only `slack_sdk.AsyncWebClient` for Slack API calls). Adding one is out of scope for this PR; tracked as a follow-up. Teams version is fixed in this PR.

File: src/chat_sdk/adapters/google_chat/adapter.py [hunk 4]
```diff
@@ -2612,53 +2612,92 @@ def _create_attachment(self, att: dict[str, Any]) -> Attachment:
...
+                    return await response.read()
+
+            # Fallback to direct URL fetch (downloadUri)
+            if url:
+                token = await adapter._get_access_token()
+                session = await adapter._get_http_session()
+                async with session.get(
+                    url,
+                    headers={"Authorization": f"Bearer {token}"},
+                ) as response:
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/google_chat/adapter.py:2705]: Addressed in 26a08bd — gchat rehydrate now validates URL via `_is_trusted_gchat_download_url` (exact `chat.googleapis.com`, `googleapis.com` + suffix `.googleapis.com`, `.googleusercontent.com`, `.google.com`) before forwarding the OAuth token. Tampered URL → raises. Test in tests/test_google_chat_adapter.py.

File: src/chat_sdk/adapters/slack/adapter.py [hunk 5]
```diff
@@ -1763,6 +1770,71 @@ async def fetch_data() -> bytes:
...
+                installation = await adapter.get_installation(team_id)
+                if installation is None:
+                    raise AuthenticationError(
+                        "slack",
+                        f"Installation not found for team {team_id}",
+                    )
+                token = installation.bot_token
+            else:
+                token = adapter._get_token()
+            return await adapter._fetch_slack_file(url, token)
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/slack/adapter.py:1976]: Addressed in 26a08bd — Slack `_fetch_slack_file` validates URL via `_is_trusted_slack_download_url` (exact `files.slack.com` + suffix `.slack.com`, `.slack-edge.com`) before forwarding the bot token. HTTPS required. Tests in tests/test_slack_webhook.py include look-alike `slack.com.attacker.tld` rejection.

File: src/chat_sdk/adapters/teams/adapter.py [hunk 6]
```diff
@@ -580,11 +581,52 @@ def _create_attachment(self, att: dict[str, Any]) -> Attachment:
...
+            type=attachment.type,
+            url=attachment.url,
+            name=attachment.name,
+            mime_type=attachment.mime_type,
+            size=attachment.size,
+            width=attachment.width,
+            height=attachment.height,
+            data=attachment.data,
+            fetch_data=self._build_teams_fetch_data(url),
+            fetch_metadata=attachment.fetch_metadata,
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/teams/adapter.py:689]: Addressed in 26a08bd — Teams `_build_teams_fetch_data` validates URL via `_is_trusted_teams_download_url` before issuing the GET. Comprehensive allowlist covering Graph/SharePoint/Bot Framework/trafficmanager. Test verifies `_get_http_session.assert_not_awaited()` on rejection (validation short-circuits before any session work).

File: src/chat_sdk/adapters/telegram/adapter.py [hunk 7]
```diff
@@ -1365,6 +1365,32 @@ def create_attachment(
...
+
+    def rehydrate_attachment(self, attachment: Attachment) -> Attachment:
+        """Reconstruct ``fetch_data`` on a deserialized Telegram attachment.
+
+        Pulls ``fileId`` from ``fetch_metadata`` and rebuilds the lazy
+        ``download_file`` closure.  Returns the attachment unchanged when
+        no file ID is present (e.g. a pre-serialized attachment that did
+        not originate from this adapter).
+        """
+        meta = attachment.fetch_metadata or {}
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/telegram/adapter.py]: Addressed in 26a08bd — `attachment.fetch_metadata or {}` replaced with `is not None` pattern.

File: src/chat_sdk/adapters/whatsapp/adapter.py [hunk 8]
```diff
@@ -633,6 +633,31 @@ def _build_media_attachment(
...
+        )
+
+    def rehydrate_attachment(self, attachment: Attachment) -> Attachment:
+        """Reconstruct ``fetch_data`` on a deserialized WhatsApp attachment.
+
+        Pulls ``mediaId`` from ``fetch_metadata`` and rebuilds the lazy
+        ``download_media`` closure.  Returns the attachment unchanged when
+        no media ID is present.
+        """
+        meta = attachment.fetch_metadata or {}
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/whatsapp/adapter.py]: Addressed in 26a08bd — same truthiness fix in whatsapp adapter.

File: src/chat_sdk/chat.py [hunk 9]
```diff
@@ -2037,52 +2038,75 @@ def _detect_mention(self, adapter: Adapter, message: Message) -> bool:
...
+            return raw  # type: ignore[return-value]
+
+        # Apply the adapter's rehydrate_attachment hook (if provided) to any
+        # attachment that lost its fetch_data closure during serialization.
+        # Matches TS: `adapter?.rehydrateAttachment?.(att)` — duck-typed so
+        # adapters that do not declare the hook (e.g. bare MockAdapter) are
+        # treated as no-ops and the attachment is left untouched.
+        rehydrate = getattr(adapter, "rehydrate_attachment", None) if adapter else None
+        if callable(rehydrate) and msg.attachments:
+            msg.attachments = [att if att.fetch_data is not None else rehydrate(att) for att in msg.attachments]
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/chat.py]: Addressed in 717a2d2 — bound the `getattr` result to explicit `Callable[[Attachment], Attachment] | None` local. Pyrefly narrows the list comprehension to `list[Attachment]`. CI blocker resolved.

File: src/chat_sdk/types.py [hunk 10]
```diff
@@ -498,6 +518,7 @@ def from_json(cls, data: dict[str, Any] | Message) -> Message:
                     size=att.get("size"),
                     width=att.get("width"),
                     height=att.get("height"),
+                    fetch_metadata=att.get("fetchMetadata") or att.get("fetch_metadata"),
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/types.py]: Addressed in c871b2b — both `from_json` and `from_json_compat` now use explicit `is not None` ternary for the dual-key lookup (catches `mime_type` too, which the earlier pass missed).

File: tests/test_slack_webhook.py [hunk 11]
```diff
@@ -510,6 +510,100 @@ def make_event(mimetype: str) -> dict:
...
+        adapter = _make_adapter(bot_token="xoxb-single")
+        rehydrated = adapter.rehydrate_attachment(
+            Attachment(
+                type="image",
+                url="https://files.slack.com/img.png",
+                fetch_metadata={"url": "https://files.slack.com/img.png"},
+            )
+        )
+        assert rehydrated.fetch_data is not None
+
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply tests/test_slack_webhook.py:615]: Addressed in 26a08bd — tests stub `_fetch_slack_file` with AsyncMock, await the closure, and assert which token path was exercised (workspace install vs static bot token). No longer passing on a stub closure.

File: tests/test_teams_adapter.py [hunk 12]
```diff
@@ -327,6 +327,62 @@ def test_edited_false_for_new(self):
...
+        assert rehydrated.fetch_data is not None
+
+    def test_rehydrate_falls_back_to_attachment_url_when_fetch_metadata_missing(self):
+        """When fetch_metadata is absent, rehydrate falls back to the attachment's top-level url."""
+        from chat_sdk.types import Attachment
+
+        adapter = _make_adapter(app_id="test-app")
+        attachment = Attachment(type="file", url="https://x.com/doc.pdf")
+        rehydrated = adapter.rehydrate_attachment(attachment)
+        assert rehydrated.fetch_data is not None
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply tests/test_teams_adapter.py]: Addressed in 26a08bd — tests stub `_get_http_session` with aiohttp-shaped mock, await `rehydrated.fetch_data()`, assert URL + returned bytes.
